### PR TITLE
kernel: Implement thread pinning support

### PR DESCRIPTION
**Pull request by Thog**
This PR adds support for 8.x thread pinning changes and implement SynchronizePreemptionState syscall.

Based on kernel 13.x reverse.

As always need heavy testing.

### CODE REVIEW

File: Ryujinx.HLE/HOS/Kernel/Process/KProcess.cs [hunk 1]
```diff
@@ -9,6 +9,7 @@
 using Ryujinx.Memory;
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Kernel/Process/KProcess.cs:12]: Not needed?

File: Ryujinx.HLE/HOS/Kernel/Threading/KThread.cs [hunk 2]
```diff
@@ -432,6 +450,21 @@ private bool ExitImpl()
             return decRef;
         }
 
+        private int GetEffectiveRunningCore()
+        {
+            int coreNumber;
+
+            for (coreNumber = 0; coreNumber < KScheduler.CpuCoresCount; coreNumber++)
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Kernel/Threading/KThread.cs:457]:
```suggestion
            for (int coreNumber = 0; coreNumber < KScheduler.CpuCoresCount; coreNumber++)
```

File: Ryujinx.HLE/HOS/Kernel/Threading/KThread.cs [hunk 3]
```diff
@@ -629,6 +708,51 @@ public KernelResult SetCoreAndAffinityMask(int newCore, long newAffinityMask)
...
+                    }
+                    else
+                    {
+                        targetThreadPinned = true;
+                    }
+                }
+
+                KernelContext.CriticalSection.Leave();
+            }
+
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Kernel/Threading/KThread.cs:754]: Double empty line.

File: Ryujinx.HLE/HOS/Kernel/Process/KProcess.cs [hunk 4]
```diff
@@ -745,9 +747,27 @@ public void SubscribeThreadEventHandlers(ARMeilleure.State.ExecutionContext cont
 
         private void InterruptHandler(object sender, EventArgs e)
         {
+            KernelContext kernelContext = KernelStatic.GetContext();
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Kernel/Process/KProcess.cs:750]: Not needed as `KProcess` already has a `KernelContext` instance inherited from `KAutoObject` called `KernelContext`.

File: Ryujinx.HLE/HOS/Kernel/SupervisorCall/SyscallHandler.cs [hunk 5]
```diff
@@ -19,7 +19,23 @@ public SyscallHandler(KernelContext context)
 
         public void SvcCall(object sender, InstExceptionEventArgs e)
         {
-            ExecutionContext context = (ExecutionContext)sender;
+            KernelContext kernelContext = KernelStatic.GetContext();
```
Comment (gdkchan) [inline Ryujinx.HLE/HOS/Kernel/SupervisorCall/SyscallHandler.cs:22]: Also not needed as there is a instance of the `KernelContext` on the class called `_context`, thus the `GetContext` function can be removed.

File: Ryujinx.HLE/HOS/Kernel/Process/KProcess.cs (hunk 4 above)
Comment (Thog) [reply Ryujinx.HLE/HOS/Kernel/Process/KProcess.cs:750]: Quite interesting that I didn't thought about that considering I use it in the same function too...

**Comment by Thog** [comment]
Fixed all comments and rebased.

**Review by gdkchan — APPROVED** [review]
lgtm, thanks. Hopefully it will fix some of the intermittent softlocks.

**Review by AcK77 — APPROVED** [review]

**Pull request merged**